### PR TITLE
Specmate Dependency Matcher

### PR DESCRIPTION
**Pull request by Daedo**
Added Dependency Match Architecture.

### CODE REVIEW

**Comment by junkerm** [comment]
please remove / ignore the bundles/.metadata files (or are they there for a reason)

File: bundles/specmate-integration-test/src/com/specmate/test/integration/NLPServiceTest.java [hunk 1]
```diff
@@ -135,7 +205,7 @@ private void checkCauseEffect(String text, ELanguage language, String expectedCa
 				new ZeroOrMoreConsumer(result, new CoveredTextMatcher(",")),
 				new ZeroOrMoreConsumer(result, new ConstituentTypeMatcher("ADVP")),
 				new ZeroOrMoreConsumer(result, new AnyMatcher(), "effect"))));
-		MatchResult matchResult = matcher.match(cons);
+		com.specmate.nlp.matcher.MatchResult matchResult = matcher.match(cons);
```
Comment (junkerm) [inline bundles/specmate-integration-test/src/com/specmate/test/integration/NLPServiceTest.java:207]: Is there are reason for the fully qualified class name?

File: bundles/specmate-nlp/src/com/specmate/nlp/dependency/DependencyData.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+import de.tudarmstadt.ukp.dkpro.core.api.segmentation.type.Token;
+import de.tudarmstadt.ukp.dkpro.core.api.syntax.type.dependency.Dependency;
+
+/**
+ * Representation for a dependency parse.
+ * 
+ * @author Dominik
+ *
+ */
+public class DependencyData {
```
Comment (junkerm) [inline bundles/specmate-nlp/src/com/specmate/nlp/dependency/DependencyData.java]: Could a better name be DependencyTree?

File: bundles/specmate-nlp/src/com/specmate/nlp/dependency/DependencyData.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+	private static String ROOT = "ROOT";
+	
+	private Map<Token, DependencyNode> dependencies;
+	private Set<Token> heads;
+	private List<TextInterval> treeFragments;
+	
+	public static DependencyData generateFromJCas(JCas jcas) {
+		DependencyData out = new DependencyData();
```
Comment (junkerm) [inline bundles/specmate-nlp/src/com/specmate/nlp/dependency/DependencyData.java]: suggestion: rename this to "result", that is a  convention that we follow

File: bundles/specmate-nlp/src/com/specmate/nlp/dependency/DependencyData.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+				out.dependencies.get(governor).addDepenency(d);
+				out.treeFragments.add(out.new TextInterval(d.getDependent()));
+			}
+		}
+		out.minimizeTreeFragments();
+		return out;
+	}
+
+	public static DependencyData getSubtree(DependencyData data, Token token) {
+		DependencyData out = new DependencyData(token);
```
Comment (junkerm) [inline bundles/specmate-nlp/src/com/specmate/nlp/dependency/DependencyData.java]: rename to result

File: bundles/specmate-nlp/src/com/specmate/nlp/dependency/DependencyData.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+		this.treeFragments = new Vector<DependencyData.TextInterval>();
+	}
+	
+	public void addSubtree(DependencyData subtree) {
+		this.dependencies.putAll(subtree.dependencies);
+		this.treeFragments.addAll(subtree.treeFragments);
+		this.minimizeTreeFragments();
+	}
+	
+	public void addSubtree(DependencyData subtree, Dependency connection) {
```
Comment (junkerm) [inline bundles/specmate-nlp/src/com/specmate/nlp/dependency/DependencyData.java]: connection --> dependency

File: bundles/specmate-nlp/src/com/specmate/nlp/dependency/matcher/MatchUtil.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+public class MatchUtil {
+	/**
+	 * Applies a list of matchers on the given dependency data object.
+	 * 
+	 * @param rules
+	 * @param data
+	 * @return A List of Results for each head of the dependency data object.
+	 */
+	public static List<MatchResult> evaluateRuleset(List<Matcher> rules, DependencyData data) {
+		Vector<MatchResult> out = new Vector<MatchResult>();
```
Comment (junkerm) [inline bundles/specmate-nlp/src/com/specmate/nlp/dependency/matcher/MatchUtil.java]: out --> result

File: bundles/specmate-nlp/src/com/specmate/nlp/dependency/matcher/MatchUtil.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+	/**
+	 * Recursively applies the list of matchers on the given dependency data object starting at the head token.
+	 * 
+	 * @param rules
+	 * @param data
+	 * @param head
+	 * @return
+	 */
+	public static MatchResult evaluateRuleset(List<Matcher> rules, DependencyData data, Token head) {
+		MatchResult out = MatchResult.unsuccessful();
```
Comment (junkerm) [inline bundles/specmate-nlp/src/com/specmate/nlp/dependency/matcher/MatchUtil.java]: out --> result

File: bundles/specmate-nlp/src/com/specmate/nlp/dependency/DependencyData.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+				begin = other;
+				end = this;
+			}
+			
+			if(begin.to == end.from) {
+				return new TextInterval(begin.from, end.to, begin.text+end.text);
+			}
+			
+			if(begin.to == end.from-1) {
+				return new TextInterval(begin.from, end.to, begin.text+" "+end.text);
```
Comment (junkerm) [inline bundles/specmate-nlp/src/com/specmate/nlp/dependency/DependencyData.java]: I wonder if this is valid, can you be sure the gap is always a space?

File: bundles/specmate-integration-test/src/com/specmate/test/integration/NLPServiceTest.java [hunk 9]
```diff
@@ -121,6 +128,69 @@ public void testMatcher() throws SpecmateException {
 
 	}
 
+	@Test
+	public void testDependencyParse() throws SpecmateException {
```
Comment (junkerm) [inline bundles/specmate-integration-test/src/com/specmate/test/integration/NLPServiceTest.java:133]: so far, the OptionMatcher is not tested here, can you modiffy your test to also use the OptionMatcher

Can you also add tests for cases where no matches occur or only parts match?

**Review by junkerm — APPROVED** [review]
Looks good, I only have minor comments. If you adress these please merge.

File: bundles/specmate-integration-test/src/com/specmate/test/integration/NLPServiceTest.java (hunk 1 above)
Comment (Daedo) [reply bundles/specmate-integration-test/src/com/specmate/test/integration/NLPServiceTest.java:207]: We currently have two differend "MatchResut" classes. This one is from a previous attempt at implementing the dependency ruleset and probably gets removed once the new version is up and running.

**Review by junkerm — APPROVED** [review]
looks good, go ahed and merge

**Pull request merged**